cql3/expr/to_restriction.cc

```diff
@@ -20,14 +20,164 @@
+                    oper);
+    }
+}
+
+::shared_ptr<restrictions::restriction> new_single_column_restriction(column_value&& lhs_col,
+                                                                      oper_t oper,
+                                                                      expression&& prepared_rhs,
+                                                                      comparison_order order,
+                                                                      const schema& schema) {
+    validate_single_column_relation(lhs_col, oper, schema, false);
+
+    ::shared_ptr<restrictions::restriction> r = ::make_shared<restrictions::single_column_restriction>(*lhs_col.col);
+    r->expression = binary_operator(std::move(lhs_col), oper, std::move(prepared_rhs), order);
+    return r;
+}
+
+::shared_ptr<restrictions::restriction> new_subscripted_column_restriction(subscript&& lhs_sub,
+                                                                           oper_t oper,
+                                                                           expression&& prepared_rhs,
+                                                                           comparison_order order,
+                                                                           const schema& schema) {
+    const column_value& sub_col = get_subscripted_column(lhs_sub);
+    validate_single_column_relation(sub_col, oper, schema, true);
+
+    ::shared_ptr<restrictions::restriction> r = ::make_shared<restrictions::single_column_restriction>(*sub_col.col);
+    r->expression = binary_operator(std::move(lhs_sub), oper, std::move(prepared_rhs));
+    return r;
+}
+
+} // anonymous namespace
+
 ::shared_ptr<restrictions::restriction> to_restriction(const expression& e,
                                                        data_dictionary::database db,
                                                        schema_ptr schema,
                                                        prepare_context& ctx) {
-    throw std::runtime_error(fmt::format("Unimplemented! {}:{}", __FILE__, __LINE__));
+    if (is<constant>(e)) {
+        throw exceptions::invalid_request_exception("Constant value by itself is not supported as a restriction yet");
+    }
+
+    if (!is<binary_operator>(e)) {
+        throw exceptions::invalid_request_exception("Restriction must be a binary operator");
+    }
+    binary_operator binop_to_prepare = as<binary_operator>(e);
+
+    // Convert single element IN relation to an EQ relation
+    if (binop_to_prepare.op == oper_t::IN && is<collection_constructor>(binop_to_prepare.rhs)) {
+        const std::vector<expression>& elements = as<collection_constructor>(binop_to_prepare.rhs).elements;
+        if (elements.size() == 1) {
+            binop_to_prepare.op = oper_t::EQ;
+            binop_to_prepare.rhs = elements[0];
+        }
+    }
+
+    binary_operator prepared_binop = prepare_binary_operator(binop_to_prepare, db, schema, ctx);
```

Comment:
For later: we can call prepare_expression() here (and remove the restriction that it must be a binop). The receiver will have type boolean, since restrictions must evaluate to booleans. This will enable `WHERE bool_column` and the mighty `WHERE TRUE`.